picamera2/remote.py

```diff
@@ -0,0 +1,429 @@
+        self._init_func = init
+        self.start()
+
+    def _child_init(self):
+        """Initialization that runs in the child process."""
+        self._syscall = CDLL(None, use_errno=True).syscall
+        self._syscall.argtypes = [c_long]
+        if self._init_func is not None:
+            self._init_func()
+
+    def run(self):
+        """Main loop of the child process."""
+        self._child_init()
+        while (request := self._capture_request()) is not None:
+            value = self._run(request)
+            self._return_request(value)
+
+    def _map_fd(self, picam2_fd: int):
+        """Maps a file descriptor from the parent process to the child process."""
+        if self._pid_fd is None:
+            self._pid_fd = os.pidfd_open(self._picam2_pid)
+        fd = self._syscall(438, c_int(self._pid_fd), c_int(picam2_fd), c_int(0))  # 438 is pidfd_getfd
+        if fd == -1:
+            errno = get_errno()
+            raise OSError(errno, os.strerror(errno))
+        return fd
+
+    def _capture_request(self):
+        """Captures a request from the parent process."""
+        if self._return_result:
+            self._return_queue.put(None)
+        self._return_result = True
+
+        msg = self._send_queue.get()
+        if msg == "CLOSE":
+            self._return_queue.put(None)
+            return None
+
+        msg._deserialize(self)
+        self._request = msg
+        return msg
+
+    def _return_request(self, return_value: Any):
+        """Returns a request to the parent process."""
+        if self._request is None:
+            raise ValueError("No request to return")
+
+        if self._request._array_ref_count > 0:
+            raise ValueError("Cannot return request while arrays are still in use")
+
+        self._request = None
+        self._return_queue.put(return_value)
+        self._return_result = False
+
+
+class RemoteMappedArray:
+    """
+    A mapped array that is created in the child process.
+
+    Args:
+        request: The RemoteRequest object
+        stream_name: The name of the stream to map
+    """
+
+    def __init__(self, request: "RemoteRequest", stream_name: str):
+        """Initializes the RemoteMappedArray."""
+        self._request = request
+        self._stream_name = stream_name
+        self._array = None
+
+    def __enter__(self):
+        self._request._array_ref_count += 1
+        self._array = self._request._arrays[self._stream_name]
+        return self
+
+    def __exit__(self, exc_type, exc_value, exc_traceback):
+        self._request._array_ref_count -= 1
+        if self._request._array_ref_count < 0:
+            raise ValueError("Array reference count is less than 0")
+        self._array = None
+
+    @property
+    def array(self):
+        """Access the array."""
+        return self._array
+
+
+class RemoteRequest:
+    """
+    A request that is sent to the child process.
+
+    Args:
+        request: The CompletedRequest object
+    """
```

Comment:
I have a feeling we might need to think about what happens if someone writes back to the underlying camera buffer. Do we need DMA buf syncs? (i.e. flush the data cache back to memory)

If this is too painful, we could make buffers read-only.